exchanges/exchange.go

```diff
@@ -588,37 +665,70 @@ func (e *Base) UpdatePairs(exchangeProducts currency.Pairs, assetType asset.Item
 		products = append(products, exchangeProducts[x])
 	}
 
-	var newPairs, removedPairs currency.Pairs
 	var updateType string
-	targetPairs := e.CurrencyPairs.GetPairs(assetType, enabled)
+	targetPairs, err := e.CurrencyPairs.GetPairs(assetType, enabled)
+	if err != nil {
+		return err
+	}
 
 	if enabled {
-		newPairs, removedPairs = targetPairs.FindDifferences(products)
 		updateType = "enabled"
 	} else {
-		newPairs, removedPairs = targetPairs.FindDifferences(products)
 		updateType = "available"
 	}
 
+	newPairs, removedPairs := targetPairs.FindDifferences(products)
 	if force || len(newPairs) > 0 || len(removedPairs) > 0 {
 		if force {
 			log.Debugf(log.ExchangeSys,
-				"%s forced update of %s [%v] pairs.", e.Name, updateType,
+				"%s forced update of %s [%v] pairs.",
+				e.Name,
+				updateType,
 				strings.ToUpper(assetType.String()))
 		} else {
 			if len(newPairs) > 0 {
 				log.Debugf(log.ExchangeSys,
-					"%s Updating pairs [%v] - New: %s.\n", e.Name,
-					strings.ToUpper(assetType.String()), newPairs)
+					"%s Updating %s pairs [%v] - Added: %s.\n",
+					e.Name,
+					updateType,
+					strings.ToUpper(assetType.String()),
+					newPairs)
 			}
 			if len(removedPairs) > 0 {
 				log.Debugf(log.ExchangeSys,
-					"%s Updating pairs [%v] - Removed: %s.\n", e.Name,
-					strings.ToUpper(assetType.String()), removedPairs)
+					"%s Updating %s pairs [%v] - Removed: %s.\n",
+					e.Name,
+					updateType,
+					strings.ToUpper(assetType.String()),
+					removedPairs)
 			}
 		}
 		e.Config.CurrencyPairs.StorePairs(assetType, products, enabled)
 		e.CurrencyPairs.StorePairs(assetType, products, enabled)
+
+		if !enabled {
+			// If available pairs are changed we will remove currency pair items
+			// that are still included in the enabled pairs list.
+			enabledPairs, _ := e.CurrencyPairs.GetPairs(assetType, true)
```

Comment:
Can you give me an example of the race issue this presents? I guess calling the function in multiple routines UpdateTradeablePairs() would cause an issue but still don't think it would produce a race due to the lock on pairsManager. As with returning the error here, all its doing is checking if there is an enabled pair not found in the available list it still returns the full enabled pair list for filtering.

Reply:
Data race in the sense that in the time between saving above and getting here, the data could change and doesn't match your expectations.